stringdtype/stringdtype/src/static_string.c

```diff
@@ -1,41 +1,76 @@
 #include "static_string.h"
 
-// allocates a new ss string of length len, filling with the contents of init
-ss *
-ssnewlen(const char *init, size_t len)
+int
+ssnewlen(const char *init, size_t len, ss *to_init)
 {
+    if ((to_init->buf != NULL) || (to_init->len != 0)) {
+        return -1;
+    }
+
     // one extra byte for null terminator
-    ss *ret = (ss *)malloc(sizeof(ss) + sizeof(char) * (len + 1));
+    char *ret_buf = (char *)malloc(sizeof(char) * (len + 1));
 
-    if (ret == NULL) {
-        return NULL;
+    if (ret_buf == NULL) {
+        return -1;
     }
 
-    ret->len = len;
+    to_init->len = len;
 
     if (len > 0) {
-        memcpy(ret->buf, init, len);
+        memcpy(ret_buf, init, len);
     }
 
-    ret->buf[len] = '\0';
+    ret_buf[len] = '\0';
+
+    to_init->buf = ret_buf;
+
+    return 0;
+}
 
-    return ret;
+void
+ssfree(ss *str)
+{
+    if (str->buf != NULL) {
+        free(str->buf);
+        str->buf = NULL;
+    }
+    str->len = 0;
 }
 
-// returns a new heap-allocated copy of input string *s*
-ss *
-ssdup(const ss *s)
+int
+ssdup(ss *in, ss *out)
 {
-    return ssnewlen(s->buf, s->len);
+    return ssnewlen(in->buf, in->len, out);
 }
 
-// returns a new, empty string of length len
-// does not do any initialization, the caller must
-// initialize and null-terminate the string
-ss *
-ssnewempty(size_t len)
+int
+ssnewemptylen(size_t num_bytes, ss *out)
 {
-    ss *ret = (ss *)malloc(sizeof(ss) + sizeof(char) * (len + 1));
-    ret->len = len;
-    return ret;
+    if (out->len != 0 || out->buf != NULL) {
+        return -1;
+    }
+
+    char *buf = (char *)malloc(sizeof(char) * (num_bytes + 1));
+
+    if (buf == NULL) {
```

Comment:
In this case is it appropriate to set a python exception? This happens when memory can't be allocated.

Reply:
I'd prefer to leave it to the responsibility of the caller to check the error code. That way this `static_string` library doesn't need to depend on `python.h` or worry about the GIL.

That said, I'm going to make it so you get back `-2` if `out` is not empty and `-1` if malloc fails.

Reply:
Sounds good.